### PR TITLE
Adapt tests to support CI and broad them to other projects

### PR DESCRIPTION
**Pull request by amurKontur**
Expand login tests to cover all key projects and fix config

## Summary by CodeRabbit

- **New Features**
    - Updated the `.env.playwright` file to include project details for improved test automation.
    - Added functionality to dynamically generate login tests for various projects using Playwright.
- **Documentation**
    - Enhanced guidelines for disaster-ninja front-end build with Docker and Vite.
    - Updated i18n instructions and translation steps in React.
    - Provided detailed information on tests, type checks, and configuration updates.
- **Refactor**
    - Restructured color mapping logic in `mcdaStyle.ts` for better rendering.
    - Improved logic for determining cell labels in `bivariateLegendUtils.ts`.
    - Enhanced type safety and function clarity in `styleGen.ts`.
- **Tests**
    - Added edge case tests for `getCellLabelByValue` function in `bivariateLegendUtils.test.ts`.
- **Chores**
    - Updated exclusions in coverage configuration in `vite.config.ts`.

### CODE REVIEW

File: .env.playwright
Comment (dqunbp) [inline .env.playwright:1]: It is better to keep this file as is; it contains local environment configurations which will be overridden by the `.env.playwright.local` file.

File: .env.playwright [hunk 1]
```diff
@@ -1 +1,23 @@
-BASE_URL=https://localhost:3000
\ No newline at end of file
+PROJECTS='{
```
Comment (dqunbp) [inline .env.playwright]: It looks like it will be more maintainable to save this JSON config to a JSON file and then read from it when the tests start.

File: .env.playwright [hunk 2]
```diff
@@ -1 +1,23 @@
...
+        "url":"https://maps.kontur.io/",
+        "app":"634f23f5-f898-4098-a8bd-09eb7c1e1ae5",
+        "title":"Smart City"
+    },
+    "oam":{
+        "url":"https://new.openaerialmap.org/",
+        "app":"1dc6fe68-8802-4672-868d-7f17943bf1c8",
+        "title":"OpenAerialMap"
+    }
+}'
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .env.playwright]: app IDs are available publicly at our websites urls' params

File: e2e/login.spec.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { test, expect } from '@playwright/test';
+
+// Get data from config about projects and convert to JS object setting TS types
+
+const projects = JSON.parse(process.env.PROJECTS!) as Record<
+  string,
+  { url: string; app: string; title: string }
+>;
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/login.spec.ts]: project variable is a global one, without it no testing is possible

File: README.md [hunk 4]
```diff
@@ -182,7 +208,29 @@ To view the app, visit https://localhost:3000 in your browser.
 Create a `.env.playwright.local` file in the root of the project with the following content:
 
 ```bash
-BASE_URL=app_base_url
```
Comment (dqunbp) [inline README.md:185]: It looks like we use absolute app URLs; it's not a base URL anymore.

If we will not use the base URL, just remove it completely.

File: e2e/login.spec.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    await emailInput.fill(process.env.TEST_EMAIL!);
+
+    // Getting password field and filling in
+    const passwordInput = page.locator('input[type="password"]');
+    await passwordInput.fill(process.env.TEST_PASSWORD!);
+
+    // Currently OAM and Atlas apps don't have cookies popups
+    if (titles[i] !== projects.oam.title && titles[i] !== projects.atlas.title) {
+      await page.getByText('Accept optional cookies').click();
+    }
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/login.spec.ts]: Here the bug is present that was found by autotest - https://kontur.fibery.io/Tasks/Task/Add-cookie-feature-to-Kontur-apps-18184. After its fix, if clause should be deleted. Just click on popup should stay

File: playwright.config.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ export default defineConfig({
   /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
   use: {
     /* Base URL to use in actions like `await page.goto('/')`. */
-    baseURL: process.env.BASE_URL,
+    baseURL: JSON.parse(process.env.PROJECTS!).atlas.url,
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts]: no, our main focus is Atlas, so we should use it as base url with majority of tests focusing on this project

File: README.md [hunk 7]
```diff
@@ -32,33 +35,39 @@ pnpm run serve
...
+
 We use i18next + gettext convertion approach in our development toolchain.
 There are several scripts in package.json:
 
 - i18n:export converts i18next to gettext
 - i18n:import converts gettext to i18next
-- i18n:update extracts new translation keys from code to common.json (```i18n.t``` or ```<Trans/>```) and also removes unused keys
+- i18n:update extracts new translation keys from code to common.json (`i18n.t` or `<Trans/>`) and also removes unused keys
 - lint:i18n:keys:identity is needed to see untranslated keys for all locales (make sure you ran i18n:import before as it compares i18next .json files)
 - i18n:gettext-sync is needed to keep all .po files synced with .pot file (marks non-compliant msgid as 'fuzzy', renames keys in .po according keys changes in .pot)
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply README.md:55]: this changes were done by prettier, not sure that we need to keep them, better get some other dev opinion

File: .env.playwright
Comment (amurKontur) [reply .env.playwright:1]: but it now supports json, not just url

File: e2e/login.spec.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    acc.apps.push(project.app);
+    acc.titles.push(project.title);
+    return acc;
+  },
+  { urls: [], apps: [], titles: [] },
+);
+
+// Create old-school loop to loop over all the projects and create a test for everyone
+
+for (let i = 0; i < urls.length; i++) {
```
Comment (dqunbp) [inline e2e/login.spec.ts]: Please iterate over the projects itself, for example `for (project of projects)`

Comment (dqunbp) [inline e2e/login.spec.ts]: Please iterate over the projects itself, for example `for (project of projects)`

File: e2e/login.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dqunbp) [reply e2e/login.spec.ts]: Please follow the coderabbit))

[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/login.spec.ts]: @coderabbitai please open an issue

File: .env.playwright (hunk 1 above)
Comment (amurKontur) [reply .env.playwright]: I tried to do it, but it seems like some issues appear than. We should better postpone until we know how to make a workable beautiful playwright config

File: README.md (hunk 4 above)
Comment (amurKontur) [reply README.md:185]: deleted

File: e2e/login.spec.ts (hunk 8 above)
Comment (amurKontur) [reply e2e/login.spec.ts]: done

Comment (amurKontur) [reply e2e/login.spec.ts]: done

File: .env.playwright (hunk 1 above)
Comment (dqunbp) [reply .env.playwright]: What kind of issues appeared?

File: e2e/login.spec.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+const projects = JSON.parse(process.env.PROJECTS!) as Record<
+  string,
+  { url: string; app: string; title: string }
+>;
+
+// Create a loop to loop over all the projects and create a test for everyone
+
+for (const project of Object.keys(projects)) {
+  test(`As User, I can login to ${projects[project].title}`, async ({ page }) => {
```
Comment (dqunbp) [inline e2e/login.spec.ts]: We can do it without `Object.keys` then accessing to the properties directly just `project.title`, etc

File: .env.playwright (hunk 1 above)
Comment (amurKontur) [reply .env.playwright]: fixed

File: e2e/login.spec.ts (hunk 9 above)
Comment (amurKontur) [reply e2e/login.spec.ts]: better way was used finally

File: .github/workflows/qa.yml [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$test_oam_url\"|g" e2e/projects-config.json
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
+
+  - name: Move to e2e directory # Move to e2e
+    run: cd e2e
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/qa.yml]: done

File: .github/workflows/run_e2e_tests.yml [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$test_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$test_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$test_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$test_oam_url\"|g" e2e/projects-config.json
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/run_e2e_tests.yml]: fixed

File: .github/workflows/run_e2e_tests.yml [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+          if [[ "${{ github.event.inputs.env }}" == "test" ]]; then
+            echo "TEST_EMAIL=${{ secrets.TEST_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.TEST_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            echo "TEST_EMAIL=${{ secrets.DEV_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.DEV_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/run_e2e_tests.yml]: fixed

File: .github/workflows/run_e2e_tests.yml [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    env:
+      INPUT_ENV: ${{ github.event.inputs.env }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Setup Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: '20'
+      - name: Install dependencies
```
Comment (dqunbp) [inline .github/workflows/run_e2e_tests.yml]: We don't need the project dependencies

File: .github/workflows/e2e_tests.yml [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$test_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$test_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$test_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$test_oam_url\"|g" e2e/projects-config.json
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/e2e_tests.yml]: fixed in new version

File: .github/workflows/e2e_tests.yml [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+          if [[ "${{ github.event.inputs.env }}" == "test" ]]; then
+            echo "TEST_EMAIL=${{ secrets.TEST_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.TEST_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            echo "TEST_EMAIL=${{ secrets.DEV_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.DEV_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/e2e_tests.yml]: fixed in new version

File: .github/workflows/e2e.yml [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$test_oam_url\"|g" e2e/projects-config.json
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
+
+      - name: Move to e2e directory # Move to e2e
+        run: cd e2e
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/e2e.yml]: fixed in new version

File: .github/workflows/e2e.yml [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$test_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$test_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$test_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$test_oam_url\"|g" e2e/projects-config.json
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/e2e.yml]: fixed in new version

File: .github/workflows/e2e.yml [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+          if [[ "${{ github.event.inputs.env }}" == "test" ]]; then
+            echo "TEST_EMAIL=${{ secrets.TEST_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.TEST_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            echo "TEST_EMAIL=${{ secrets.DEV_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.DEV_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/e2e.yml]: fixed in new version

File: .github/workflows/qa.yml [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$test_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$test_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$test_oam_url\"|g" e2e/projects-config.json
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
+
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/qa.yml]: fixed in new version

File: .github/workflows/qa.yml [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+          if [[ "${{ github.event.inputs.env }}" == "test" ]]; then
+            echo "TEST_EMAIL=${{ secrets.TEST_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.TEST_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            echo "TEST_EMAIL=${{ secrets.DEV_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.DEV_PASSWORD }}" >> .env.playwright.local
+          elif [[ "${{ github.event.inputs.env }}" == "prod" ]]; then
+            echo "TEST_EMAIL=${{ secrets.PROD_EMAIL }}" >> .env.playwright.local
+            echo "TEST_PASSWORD=${{ secrets.PROD_PASSWORD }}" >> .env.playwright.local
+          fi
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/qa.yml]: fixed in new version

File: .github/workflows/qa.yml [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+          elif [[ "${{ github.event.inputs.env }}" == "dev" ]]; then
+            sed -i "s|\"url\": \"$prod_atlas_url\"|\"url\": \"$dev_atlas_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_disaster_ninja_url\"|\"url\": \"$dev_disaster_ninja_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_smart_city_url\"|\"url\": \"$dev_smart_city_url\"|g" e2e/projects-config.json
+            sed -i "s|\"url\": \"$prod_oam_url\"|\"url\": \"$dev_oam_url\"|g" e2e/projects-config.json
+          fi
+
+      - name: Move to e2e directory # Move to e2e
+        run: cd e2e
+
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply .github/workflows/qa.yml]: fixed in new version

**Review by dqunbp — APPROVED** [review]
